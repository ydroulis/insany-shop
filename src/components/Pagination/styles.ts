"use client"
import styled from 'styled-components'

export const Container = styled.div`
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    gap: 21px;

    margin-top: 3.188rem;

    .current {
        border: 1px solid #A212DF !important;
        color: #A212DF !important;
    }
`

export const Buttons = styled.button`
    width: 32px;
    height: 32px;
    border-radius: 8px;

    display: flex;
    justify-content: center;
    align-items: center;

    cursor: pointer;
    border: none;
    background-color: #E9E9F0;

    color: #737380;
    font-size: 1rem;

    &:hover {
        opacity: 0.6;
    }

    &:disabled {
        opacity: 0.5;
        cursor: not-allowed;
    }
`