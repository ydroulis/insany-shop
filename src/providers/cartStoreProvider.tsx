'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type CartStore, createCartStore } from '../stores/cartStore'

export type CartStoreApi = ReturnType<typeof createCartStore>

export const CartStoreContext = createContext<CartStoreApi | undefined>(
    undefined,
)

export interface CartStoreProviderProps {
    children: ReactNode
}

export const CartStoreProvider = ({
    children,
}: CartStoreProviderProps) => {
    const storeRef = useRef<CartStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createCartStore()
    }

    return (
        <CartStoreContext.Provider value={storeRef.current}>
            {children}
        </CartStoreContext.Provider>
    )
}

export const useCartStore = <T,>(
    selector: (store: CartStore) => T,
): T => {
    const cartStoreContext = useContext(CartStoreContext)

    if (!cartStoreContext) {
        throw new Error(`useStore must be used within CartStoreProvider`)
    }

    return useStore(cartStoreContext, selector)
}
