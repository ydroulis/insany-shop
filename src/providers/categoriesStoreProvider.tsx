'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type CategoriesStore, createCategoriesStore } from '../stores/categoriesStore'

export type CategoriesStoreApi = ReturnType<typeof createCategoriesStore>

export const CategoriesStoreContext = createContext<CategoriesStoreApi | undefined>(
    undefined,
)

export interface CategoriesStoreProviderProps {
    children: ReactNode
}

export const CategoriesStoreProvider = ({
    children,
}: CategoriesStoreProviderProps) => {
    const storeRef = useRef<CategoriesStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createCategoriesStore()
    }

    return (
        <CategoriesStoreContext.Provider value={storeRef.current}>
            {children}
        </CategoriesStoreContext.Provider>
    )
}

export const useCategoriesStore = <T,>(
    selector: (store: CategoriesStore) => T,
): T => {
    const categoriesStoreContext = useContext(CategoriesStoreContext)

    if (!categoriesStoreContext) {
        throw new Error(`useStore must be used within CategoriesStoreProvider`)
    }

    return useStore(categoriesStoreContext, selector)
}
