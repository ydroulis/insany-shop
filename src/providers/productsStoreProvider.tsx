'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type ProductsStore, createProductsStore } from '../stores/productsStore'

export type ProductsStoreApi = ReturnType<typeof createProductsStore>

export const ProductsStoreContext = createContext<ProductsStoreApi | undefined>(
    undefined,
)

export interface ProductsStoreProviderProps {
    children: ReactNode
}

export const ProductsStoreProvider = ({
    children,
}: ProductsStoreProviderProps) => {
    const storeRef = useRef<ProductsStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createProductsStore()
    }

    return (
        <ProductsStoreContext.Provider value={storeRef.current}>
            {children}
        </ProductsStoreContext.Provider>
    )
}

export const useProductsStore = <T,>(
    selector: (store: ProductsStore) => T,
): T => {
    const productsStoreContext = useContext(ProductsStoreContext)

    if (!productsStoreContext) {
        throw new Error(`useStore must be used within ProductsStoreProvider`)
    }

    return useStore(productsStoreContext, selector)
}
