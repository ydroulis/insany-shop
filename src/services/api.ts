import axios from "axios";

export const api = axios.create({
    baseURL: "https://api.insany.co/api",
    headers: {
        "Content-Type": "application/json",
    },
});