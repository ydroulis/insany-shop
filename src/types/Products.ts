export type Products = Product[]

export type Product = {
    id: number,
    name: string,
    description: string,
    price: number,
    image: string,
    category: string,
    stock: number,
    rating: number,
    brand?: string

}