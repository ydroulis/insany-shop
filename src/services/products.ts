import { Product, ProductsResponse } from "@/types/Products";
import { api } from "./api";

export const getProducts = async (
    page?: number,
    limit?: number,
    category?: string
): Promise<ProductsResponse> => {
    const params: Record<string, string | number> = {};
    if (page) params.page = page;
    if (limit) params.limit = limit;
    if (category) params.category = category;

    const response = await api.get<ProductsResponse>("/products", { params });
    return response.data;
};

export const getProductById = async (id: number): Promise<Product> => {
    const response = await api.get<{ product: Product }>(`/products/${id}`);
    return response.data.product;
};
