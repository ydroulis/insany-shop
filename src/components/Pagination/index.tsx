import React from 'react';

import * as S from './styles';
import { useProductsStore } from '../../providers/productsStoreProvider';
import { getProducts } from '../../services/products';
import { IoIosArrowBack } from "react-icons/io";
import { IoIosArrowForward } from "react-icons/io";

const Pagination: React.FC = () => {
    const { pagination, setProducts, setPagination } = useProductsStore((state => state));

    const handleNextPage = async () => {
        const res = await getProducts(pagination.currentPage + 1, 6);
        setPagination(res.pagination);
        setProducts(res.products);
    };

    const handlePreviousPage = async () => {
        const res = await getProducts(pagination.currentPage - 1, 6);
        setPagination(res.pagination);
        setProducts(res.products);
    };

    const handlePageClick = async (page: number) => {
        const res = await getProducts(page, 6);
        setPagination(res.pagination);
        setProducts(res.products);
    };

    const pagesArray = new Array(pagination.totalPages).fill(0).map((_, index) => index + 1);
    return (
        <S.Container>
            <S.Buttons disabled={!pagination.hasPreviousPage} onClick={() => handlePreviousPage()}>
                <IoIosArrowBack size={16} color='#737380' />
            </S.Buttons>
            {pagesArray.map(page => {
                const className = page === pagination.currentPage ? 'current' : '';
                return <S.Buttons className={className} key={page} onClick={() => handlePageClick(page)}>{page}</S.Buttons>
            })}
            <S.Buttons disabled={!pagination.hasNextPage} onClick={() => handleNextPage()}>
                <IoIosArrowForward size={16} color='#737380' />
            </S.Buttons>
        </S.Container>
    );
}

export default Pagination;