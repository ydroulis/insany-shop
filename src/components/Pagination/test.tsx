import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import Pagination from "./index";
import { useProductsStore } from "../../providers/productsStoreProvider";
import { getProducts } from "../../services/products";
import { ProductsResponse } from "@/types/Products";

jest.mock("../../providers/productsStoreProvider");
jest.mock("../../services/products");

const mockedUseProductsStore = useProductsStore as jest.MockedFunction<typeof useProductsStore>;
const mockedGetProducts = getProducts as jest.MockedFunction<typeof getProducts>;

describe("Pagination", () => {
    const setProducts = jest.fn();
    const setPagination = jest.fn();

    const makePagination = (overrides?: Partial<ProductsResponse["pagination"]>): ProductsResponse["pagination"] => ({
        currentPage: 1,
        totalPages: 3,
        hasPreviousPage: false,
        hasNextPage: true,
        totalProducts: 3,
        ...overrides
    });

    const mockStore = (pagination: ProductsResponse["pagination"]) => {
        mockedUseProductsStore.mockReturnValue({
            pagination,
            setProducts,
            setPagination,
        });
    };

    const makeResponse = (pagination: ProductsResponse["pagination"]): ProductsResponse => ({
        products: [
            {
                id: 1, name: "Produto 1",
                description: "",
                price: 0,
                image: "",
                category: undefined,
                stock: 0,
                rating: 0
            },
            {
                id: 2, name: "Produto 2",
                description: "",
                price: 0,
                image: "",
                category: undefined,
                stock: 0,
                rating: 0
            },
            {
                id: 3, name: "Produto 3",
                description: "",
                price: 0,
                image: "",
                category: undefined,
                stock: 0,
                rating: 0
            },
        ],
        pagination,
    });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it("should render properly page buttons", () => {
        mockStore(makePagination());
        render(<Pagination />);

        expect(screen.getAllByRole("button", { name: "" })).toHaveLength(2)
        expect(screen.getByText("1")).toHaveClass("current");
        expect(screen.getByText("2")).toBeInTheDocument();
        expect(screen.getByText("3")).toBeInTheDocument();
    });

    it("should call handleNextPage properly", async () => {
        const pagination = makePagination({ currentPage: 1, hasNextPage: true });
        mockStore(pagination);

        mockedGetProducts.mockResolvedValueOnce(makeResponse(makePagination({ currentPage: 2 })));

        render(<Pagination />);
        const nextButton = screen.getAllByRole("button").pop()!;
        fireEvent.click(nextButton);

        await waitFor(() => {
            expect(mockedGetProducts).toHaveBeenCalledWith(2, 6);
            expect(setPagination).toHaveBeenCalled();
            expect(setProducts).toHaveBeenCalled();
        });
    });

    it("should call handlePreviousPage properly", async () => {
        const pagination = makePagination({ currentPage: 2, hasPreviousPage: true });
        mockStore(pagination);

        mockedGetProducts.mockResolvedValueOnce(makeResponse(makePagination({ currentPage: 1 })));

        render(<Pagination />);
        const prevButton = screen.getAllByRole("button")[0];
        fireEvent.click(prevButton);

        await waitFor(() => {
            expect(mockedGetProducts).toHaveBeenCalledWith(1, 6);
            expect(setPagination).toHaveBeenCalled();
            expect(setProducts).toHaveBeenCalled();
        });
    });

    it("should call handlePageClick properly when click on number", async () => {
        const pagination = makePagination({ currentPage: 1 });
        mockStore(pagination);

        mockedGetProducts.mockResolvedValueOnce(makeResponse(makePagination({ currentPage: 3 })));

        render(<Pagination />);
        fireEvent.click(screen.getByText("3"));

        await waitFor(() => {
            expect(mockedGetProducts).toHaveBeenCalledWith(3, 6);
            expect(setPagination).toHaveBeenCalled();
            expect(setProducts).toHaveBeenCalled();
        });
    });

    it("should disable next when there is no next page", () => {
        const pagination = makePagination({ hasNextPage: false });
        mockStore(pagination);

        render(<Pagination />);
        const nextButton = screen.getAllByRole("button").pop()!;
        expect(nextButton).toBeDisabled();
    });

    it("should disable previous when there is no previous page", () => {
        const pagination = makePagination({ hasPreviousPage: false });
        mockStore(pagination);

        render(<Pagination />);
        const prevButton = screen.getAllByRole("button")[0];
        expect(prevButton).toBeDisabled();
    });

    it('shoult match snapshot', () => {
        const pagination = makePagination({ hasNextPage: true });
        mockStore(pagination);
        const { container } = render(<Pagination />);
        expect(container).toMatchSnapshot();
    });
});
