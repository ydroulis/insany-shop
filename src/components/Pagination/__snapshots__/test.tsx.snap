// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Pagination shoult match snapshot 1`] = `
<div>
  <div
    class="sc-bRKDuR ilKUNf"
  >
    <button
      class="sc-hvigdm fNbYBE"
      disabled=""
    >
      <svg
        color="#737380"
        fill="currentColor"
        height="16"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(115, 115, 128);"
        viewBox="0 0 512 512"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
        />
      </svg>
    </button>
    <button
      class="sc-hvigdm fNbYBE current"
    >
      1
    </button>
    <button
      class="sc-hvigdm fNbYBE"
    >
      2
    </button>
    <button
      class="sc-hvigdm fNbYBE"
    >
      3
    </button>
    <button
      class="sc-hvigdm fNbYBE"
    >
      <svg
        color="#737380"
        fill="currentColor"
        height="16"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(115, 115, 128);"
        viewBox="0 0 512 512"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M294.1 256L167 129c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.3 34 0L345 239c9.1 9.1 9.3 23.7.7 33.1L201.1 417c-4.7 4.7-10.9 7-17 7s-12.3-2.3-17-7c-9.4-9.4-9.4-24.6 0-33.9l127-127.1z"
        />
      </svg>
    </button>
  </div>
</div>
`;
